File: ployglots/mta-ployglot-pkg-figma-api/packages/py/figma_api/_retry.py
"""Figma-sensible retry defaults + user override with a
``force_overwrite`` feature flag.

The underlying ``polyglot-fetch-http-client`` ``AsyncClient`` ships a
retry loop — this module just hands it a config pre-tuned for Figma.

Consumer contract mirrors the ts twin::

    # 1. Default — Figma presets take effect.
    FigmaClient(token=...)

    # 2. Disable inner retry.
    FigmaClient(token=..., retry=False)

    # 3. Tweak selectively (merges on top of Figma defaults).
    FigmaClient(token=..., retry={"max_attempts": 5})

    # 4. force_overwrite = replace defaults verbatim.
    FigmaClient(
        token=...,
        retry={"max_attempts": 1, "retry_on_status": frozenset({503})},
        force_overwrite_retry=True,
    )
"""

from __future__ import annotations

from typing import Any

from fetch_http_client import RetryConfig

# Figma's typical rate-limit window is 1s; 429 + 5xx are retried.
FIGMA_DEFAULT_RETRY: dict[str, Any] = {
    "max_attempts": 3,
    "base_delay": 1.0,
    "max_delay": 30.0,
    "multiplier": 2.0,
    "retry_on_status": frozenset({429, 500, 502, 503, 504}),
}


def build_figma_retry_config(
    input_value: dict[str, Any] | bool | None,
    *,
    force_overwrite: bool = False,
) -> RetryConfig | None:
    """Compile a user retry input into a fetch-http-client ``RetryConfig``.

    Returns ``None`` when retry should be disabled (input is ``False``
    or ``None``). Otherwise returns a populated ``RetryConfig``, merged
    from Figma defaults unless ``force_overwrite`` is true.
    """
    # Explicit disable.
    if input_value is False or input_value is None:
        return None

    # No-input shortcut handled by caller passing {}.
    if input_value is True:
        return RetryConfig(**FIGMA_DEFAULT_RETRY)

    if not isinstance(input_value, dict):
        raise TypeError(
            f"retry must be dict, False, or None — got {type(input_value).__name__}"
        )

    if force_overwrite:
        return RetryConfig(**input_value)

    merged: dict[str, Any] = {**FIGMA_DEFAULT_RETRY, **input_value}
    # Preserve frozen sets — dict spread keeps the user's if they passed one.
    return RetryConfig(**merged)

File: ployglots/mta-ployglot-pkg-figma-api/packages/ts/src/retry.ts
// @ts-nocheck
/**
 * Figma-sensible retry defaults + user override with a `forceOverwrite`
 * feature flag.
 *
 * The underlying `@polyglot/fetch-http-client` AsyncClient already
 * ships a generic retry loop — this module just hands it a config
 * pre-tuned for Figma's behavior:
 *
 *   - Retry on 429 + 5xx (Figma rate-limits on 429 with Retry-After).
 *   - 3 attempts total, exponential 1s → 30s with FULL jitter.
 *   - Respect the `Retry-After` header.
 *   - Retry on transport-layer errors (ECONNRESET, etc).
 *
 * Consumer contract:
 *
 *   // 1. Default — Figma presets take effect.
 *   new FigmaClient({ token });
 *
 *   // 2. Disable inner retry (plan "Mode B" — outer AsyncClient owns it).
 *   new FigmaClient({ token, retry: false });
 *
 *   // 3. Tweak selectively (merges on top of the Figma defaults).
 *   new FigmaClient({ token, retry: { maxRetries: 5 } });
 *
 *   // 4. forceOverwrite = replace defaults verbatim (no merge).
 *   new FigmaClient({
 *     token,
 *     retry: { maxRetries: 1, retryOnStatus: [503] },
 *     forceOverwriteRetry: true,
 *   });
 */

import type { RetryConfig } from '@polyglot/fetch-http-client';

export type FigmaRetryInput = Partial<RetryConfig> | false | null | undefined;

/**
 * Figma-sensible retry defaults. Derived from:
 *   - Figma's documented rate-limiting strategy (429 + Retry-After)
 *   - Typical 5xx transient patterns (502/503/504)
 */
export const FIGMA_DEFAULT_RETRY: Partial<RetryConfig> = {
  maxRetries: 3,
  retryDelay: 1000,
  retryBackoff: 2,
  maxRetryDelay: 30_000,
  retryOnStatus: [429, 500, 502, 503, 504],
  retryOnException: true,
  respectRetryAfter: true,
};

export interface BuildFigmaRetryOptions {
  /**
   * When `true`, user retry options REPLACE the Figma defaults verbatim
   * (no merge). When `false` or omitted, user options are merged on
   * top of `FIGMA_DEFAULT_RETRY` so callers can tweak just the knobs
   * they care about.
   *
   * Named `forceOverwrite` to signal: you're opting out of Figma's
   * opinionated defaults. Use only if you have a reason (e.g. testing,
   * or a non-standard Figma deployment).
   */
  forceOverwrite?: boolean;
}

/**
 * Build the retry config that should be passed to the underlying
 * fetch-http-client AsyncClient.
 *
 * Returns `false` to mean "disable retry", `undefined` to mean
 * "AsyncClient default" (no retry), or a `Partial<RetryConfig>` to
 * enable retry with those knobs.
 */
export function buildFigmaRetryConfig(
  input: FigmaRetryInput,
  options: BuildFigmaRetryOptions = {},
): Partial<RetryConfig> | false {
  // Explicit disable.
  if (input === false || input === null) return false;

  // No input → use Figma defaults.
  if (input === undefined) return { ...FIGMA_DEFAULT_RETRY };

  // forceOverwrite → user wins verbatim, no merge with Figma defaults.
  if (options.forceOverwrite === true) return { ...input };

  // Default path → merge user on top of Figma defaults.
  return {
    ...FIGMA_DEFAULT_RETRY,
    ...input,
    // retryOnStatus: if the caller passed one, take it; otherwise keep Figma's.
    retryOnStatus: input.retryOnStatus ?? FIGMA_DEFAULT_RETRY.retryOnStatus,
  };
}
